### PR TITLE
feat: Convert a/mod.rs into a.rs

### LINKED ISSUE #10143: feature request: Promote / Downgrade module
IntelliJ's Rust plugin supports:
- Promote Module to directory, which means make a `.rs` file module into a directory style module with the same name.
- Downgrade Module to file, which means make a directory style module with only a `mod.rs` inside into a file module.

I use these feature a lot, but when migrating to vscode and rust-analyzer, I found there're no such things.

So is it possible for rust-analyzer to support these?

(I'd love to have a try to implment this feature if you think this is a good idea.)

### PR DESCRIPTION
**Pull request by longfangsong**
This is the reverse operation of #10211.

demo

Close #10143.

### CODE REVIEW

**Comment by lnicola** [comment]
LGTM, but one nit: would you mind changing the label to `Convert X to Y` or `Turn X into Y`? It sounds a bit better.

File: crates/ide_assists/src/handlers/move_from_mod_rs.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    if trimmed_selected_range != trimmed_file_range {
+        cov_mark::hit!(not_all_selected);
+        return None;
+    }
+
+    let target = TextRange::new(
+        source_file.syntax().text_range().start(),
+        source_file.syntax().text_range().end(),
+    );
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/move_from_mod_rs.rs:42]: This is just `source_file.syntax().text_range()`

File: crates/ide_assists/src/handlers/move_from_mod_rs.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    base_db::AnchoredPathBuf,
+};
+use syntax::{ast, AstNode, TextRange};
+
+use crate::assist_context::{AssistContext, Assists};
+use crate::utils::trimmed_text_range;
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/move_from_mod_rs.rs:8]: Nit:
```suggestion
use crate::{assist_context::{AssistContext, Assists}, utils::trimmed_text_range};
```
We use deep nested import style

File: crates/vfs/src/vfs_path.rs [hunk 3]
```diff
@@ -359,6 +359,9 @@ impl VirtualPath {
             }
             path = &path["../".len()..]
         }
+        while path.starts_with("./") {
+            path = &path["./".len()..]
+        }
```
Comment (Veykril) [inline crates/vfs/src/vfs_path.rs:364]:
```suggestion
        path = path = path.trim_start_matches("./");
```

**Comment by Veykril** [comment]
Thanks!
bors r+

**Pull request merged**